ports/raspberrypi/common-hal/nvm/ByteArray.c

```diff
@@ -0,0 +1,74 @@
+/*
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2020 microDev
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "common-hal/nvm/ByteArray.h"
+
+#include <string.h>
+
+#include "py/runtime.h"
+#include "src/rp2_common/hardware_flash/include/hardware/flash.h"
+
+extern uint32_t __flash_binary_start;
+
+uint32_t common_hal_nvm_bytearray_get_length(nvm_bytearray_obj_t* self) {
+    return self->len;
+}
+
+static void write_page(uint32_t page_addr, uint32_t offset, uint32_t len, uint8_t* bytes) {
+    // Write a whole page to flash, buffering it first and then erasing and rewriting
+    // it since we can only clear a whole page at a time.
+    if (offset == 0 && len == FLASH_PAGE_SIZE) {
```

Comment:
I think you want to use FLASH_SECTOR_SIZE (4k) instead of FLASH_PAGE_SIZE (256). There are checks for the 4k alignment here: https://github.com/raspberrypi/pico-sdk/blob/master/src/rp2_common/hardware_flash/flash.c#L69-L70

Reply:
@tannewt Thanks for the review. 

`flash_range_program()` accepts 256 aligned write... `flash_range_erase()` can only erase in 4k alignment which I think is the cause of a bug I am facing... writing at the same address twice is not working. The stored value turns to `0` when a second write attempt is made. 

Buffering in the whole sector... then erasing it and writing might fix the bug but this is highly in-efficient and buffering in the whole sector is also causing a crash.

Reply:
> Buffering in the whole sector... then erasing it and writing might fix the bug but this is highly in-efficient and buffering in the whole sector is also causing a crash.

Yes, it is inefficient. It's also the only way to do it. Programming flash only sets 0s. It cannot change a 0 to a 1. Erases reset everything back to 1. So, you always need to erase and then program in the smallest erase block size.

Also, be aware that the SDK checks may only be active during a debug build. So if it's "working" with OPT, it may still break with DEBUG=1. Dave found an issue with my code this way.

Reply:
I added a check to determine if a write procedure is a rewrite... if it is than an erase and sector write is done else page write.

> Also, be aware that the SDK checks may only be active during a debug build. So if it's "working" with OPT, it may still break with DEBUG=1. Dave found an issue with my code this way.

ya... my usual workflow is doing debug builds while in development and a test with non-debug when performing commit.

Reply:
I don't think this will work. You *always* need to erase before programming unless you know the contents are already all 1s. Otherwise the old 0s will remain even if you try to write a 1.

Reply:
yup... that is what the check does... it ensures that the range being written has all 1s.

Reply:
Oh, I see. You do write sector if any value is not 0xff. But regular program otherwise.